README.md
# Yandex-S3

Asyncio compatible SDK for Yandex Object Storage inspired by [aioaws](https://github.com/samuelcolvin/aioaws).

This library does not depend on boto, boto3 or any of the other bloated, opaque and mind thumbing AWS SDKs. Instead, it
is written from scratch to provide clean, secure and easily debuggable access to AWS services I want to use.

It currently supports:
* **S3** - list, delete, recursive delete, generating signed upload URLs, generating signed download URLs
* [AWS Signature Version 4](https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-auth-using-authorization-header.html)
  authentication for any AWS service (this is the only clean & modern implementation of AWS4 I know of in python, see
  [`core.py`](https://github.com/mrslow/yandex-s3/blob/master/yandex-s3/core.py#L114-L175))

The only dependencies of **yandex-s3**, are:
* **aiofiles** - for asynchronous reading of files
* **cryptography** - for verifying SNS signatures
* **httpx** - for HTTP requests
* **pydantic** - for validating responses

## Install

```bash
pip install yandex-s3
```

## S3 Usage


```py
import asyncio
# requires `pip install yandex-s3`
from s3 import S3Client

# requires `pip install devtools`
from devtools import debug

async def s3_demo():
    s3 = S3Client(
            access_key='<access key>', 
            secret_key='<secret key>', 
            region='<region>', 
            s3_bucket='test-bucket'
    )

    # upload a file:
    await s3.upload('dir/subdir/dummy.txt', b'Dummy')

    # list all files in a bucket
    files = [f async for f in s3.list()]
    debug(files)
    """
    [
        S3File(
            key='dir/subdir/dummy.txt',
            last_modified=datetime.datetime(...),
            size=5,
            e_tag='...',
            storage_class='STANDARD',
        ),
    ] (list) len=1
    """
    # list all files with a given prefix in a bucket
    files = [f async for f in s3.list('dir/subdir/')]
    debug(files)

    # # delete a file
    # await s3.delete('dir/subdir/dummy.txt')
    # # delete two files
    # await s3.delete('dir/subdir/file1.txt', 'dir/subdir/file2.txt')
    # delete recursively based on a prefix
    await s3.delete_recursive('dir/subdir')

    # generate an upload link suitable for sending to a borwser to enabled
    # secure direct file upload (see below)
    upload_data = s3.signed_upload_url(
        path='dir/subdir',
        filename='dummy.txt',
        content_type='text/plain',
        size=123,
    )
    debug(upload_data)
    """
    {
        'url': 'https://test-bucket.storage.yandexcloud.net/',
        'fields': {
            'Key': 'dir/subdir/dummy.txt',
            'Content-Type': 'text/plain',
            'AWSAccessKeyId': '...',
            'Content-Disposition': 'attachment; filename="dummy.txt"',
            'Policy': ('...'),
            'Signature': '...',
        },
    } (dict) len=2
    """

    # generate a temporary link to allow yourself or a client to download a file
    download_url = s3.signed_download_url('dir/subdir/dummy.txt', max_age=60)
    print(download_url)
    #> https://test-bucket.storage.yandexcloud.net/dir/subdir/dummy.txt?AWSAccessKeyId=...&Signature=...&Expires=...
    
    # download file
    content = await s3.download('dir/subdir/dummy.txt')
    print(content)
    #> b'Dummy'
    

async def main():
    await s3_demo()

asyncio.run(main())
```


CHANGELOG.md
# Changelog

#### v0.1.1
- Исправлен README.md. Теперь примеры рабочие.
- Убран python3.7 из матрицы допустимых версий python.

#### v0.1.0
- Реализован базовый функционал.

tests/test_s3.py
import pytest

from datetime import datetime
from uuid import uuid4

from s3.core import RequestError

base_url = 'https://test-bucket.storage.yandexcloud.net'

list_data = (
    b'<?xml version="1.0" encoding="UTF-8"?>\n'
    b'<ListBucketResult xmlns="http://s3.amazonaws.com/doc/2006-03-01/">\n'
    b'  <Name>flk-test-bucket</Name>\n'
    b'  <Prefix></Prefix>\n'
    b'  <MaxKeys>1000</MaxKeys>\n'
    b'  <IsTruncated>false</IsTruncated>\n'
    b'  <EncodingType>url</EncodingType>\n'
    b'  <Contents>\n'
    b'    <Key>dir/subdir/dummy.txt</Key>\n'
    b'    <LastModified>2021-08-19T08:39:21.409Z</LastModified>\n'
    b'    <Owner>\n'
    b'      <ID>ajef9a0uck81dkvnc8jf</ID>\n'
    b'      <DisplayName>ajef9a0uck81dkvnc8jf</DisplayName>\n'
    b'    </Owner>\n'
    b'    <ETag>&#34;bcf036b6f33e182d4705f4f5b1af13ac&#34;</ETag>\n'
    b'    <Size>5</Size>\n'
    b'    <StorageClass>STANDARD</StorageClass>\n'
    b'  </Contents>\n'
    b'  <Marker></Marker>\n'
    b'</ListBucketResult>\n'
)

delete_data = (
    b'<?xml version="1.0" encoding="UTF-8"?>\n'
    b'<DeleteResult xmlns="http://s3.amazonaws.com/doc/2006-03-01/">\n'
    b'  <Deleted>\n'
    b'    <Key>dir/subdir/dummy.txt</Key>\n'
    b'    <VersionId>null</VersionId>\n'
    b'  </Deleted>\n'
    b'</DeleteResult>'
)

get_data = b'Dummy'

base_headers = {
    'Server': 'nginx',
    'Connection': 'keep-alive',
    'Keep-Alive': 'timeout=60',
}

upload_headers = {
    'Date': datetime.now().strftime('%a, %d %b %Y %H:%M:%S GMT'),
    'Transfer-Encoding': 'chunked',
    'Etag': f'"{uuid4().hex}"',
    'X-Amz-Request-Id': uuid4().hex[:16],
    'X-Amz-Version-Id': None
}.update(base_headers)

list_headers = {
    'Date': datetime.now().strftime('%a, %d %b %Y %H:%M:%S GMT'),
    'Content-Type': 'application/xml; charset=UTF-8',
    'Content-Length': len(list_data),
    'X-Amz-Request-Id': uuid4().hex[:16],
}.update(base_headers)

delete_headers = {
    'Date': datetime.now().strftime('%a, %d %b %Y %H:%M:%S GMT'),
    'Content-Type': 'application/xml; charset=UTF-8',
    'Content-Length': len(delete_data),
    'X-Amz-Request-Id': uuid4().hex[:16],
}.update(base_headers)

get_headers = {
    'Date': datetime.now().strftime('%a, %d %b %Y %H:%M:%S GMT'),
    'Content-Type': 'application/octet-stream',
    'Content-Length': f'{len(get_data)}',
    'Accept-Ranges': 'bytes',
    'Etag': f'"{uuid4().hex}"',
    'Last-Modified': datetime.now().strftime('%a, %d %b %Y %H:%M:%S GMT'),
    'X-Amz-Request-Id': uuid4().hex[:16],
    'X-Amz-Version-Id': None
}.update(base_headers)


@pytest.mark.asyncio
async def test_upload_object(s3_client, httpx_mock):
    httpx_mock.add_response(
        url=base_url,
        method='POST',
        status_code=204,
        headers=upload_headers,
        data=b'')

    try:
        await s3_client.upload('dummy.txt', b'Dummy')
    except RequestError as exc:
        pytest.fail(exc)


@ pytest.mark.asyncio
async def test_download_object(s3_client, httpx_mock):
    httpx_mock.add_response(
        url=s3_client.signed_download_url('dir/subdir/dummy.txt'),
        method='GET',
        status_code=200,
        headers=get_headers,
        data=get_data)

    r = await s3_client.download('dir/subdir/dummy.txt')
    assert r == b'Dummy'


@pytest.mark.asyncio
async def test_list_objects(s3_client, httpx_mock):
    httpx_mock.add_response(
        url=f'{base_url}?list-type=2&prefix=dir/subdir&continuation-token=',
        method='GET',
        status_code=200,
        headers=list_headers,
        data=list_data)

    objects = [elem async for elem in s3_client.list(prefix='dir/subdir')]
    assert len(objects) == 1
    assert objects[0].key == 'dir/subdir/dummy.txt'


@pytest.mark.asyncio
async def test_delete_objects(s3_client, httpx_mock):
    httpx_mock.add_response(
        url=f'{base_url}?delete=1',
        method='POST',
        status_code=200,
        headers=delete_headers,
        data=delete_data)

    obj = await s3_client.delete('dir/subdir/dummy.txt')
    assert len(obj) == 1
    assert obj[0] == 'dir/subdir/dummy.txt'


tests/conftest.py
import pytest

from s3.s3 import S3Client


@pytest.fixture
async def s3_client():
    client = S3Client(
        access_key='testing',
        secret_key='testing',
        region='ru-central1',
        s3_bucket='test-bucket'
    )
    yield client
    await client.close()


s3/core.py
import hashlib
import hmac
import logging
from base64 import b64encode
from binascii import hexlify
from functools import reduce
from typing import TYPE_CHECKING, Any, Dict, Literal, Optional

from httpx import URL, AsyncClient, Response

from ._utils import get_config_attr, utcnow

if TYPE_CHECKING:
    from ._types import BaseConfigProtocol

__all__ = 'YandexS3Client', 'RequestError'
logger = logging.getLogger('s3.core')

_AWS_AUTH_REQUEST = 'aws4_request'
_CONTENT_TYPE = 'application/x-www-form-urlencoded'
_CANONICAL_REQUEST = """\
{method}
{path}
{query}
{canonical_headers}
{signed_headers}
{payload_sha256_hash}"""
_AUTH_ALGORITHM = 'AWS4-HMAC-SHA256'
_CREDENTIAL_SCOPE = '{date_stamp}/{region}/{service}/{auth_request}'
_STRING_TO_SIGN = """\
{algorithm}
{x_amz_date}
{credential_scope}
{canonical_request_hash}"""
_AUTH_HEADER = ('{algorithm} Credential={access_key}/{credential_scope},'
                'SignedHeaders={signed_headers},Signature={signature}')


class YandexS3Client:
    """
    HTTP client for Yandex Object Storage with authentication
    """

    def __init__(self, config: 'BaseConfigProtocol'):
        self.client = AsyncClient(timeout=30)
        self.access_key = get_config_attr(config, 'access_key')
        self.secret_key = get_config_attr(config, 'secret_key')
        self.region = get_config_attr(config, 'region')
        bucket = get_config_attr(config, 's3_bucket')
        self.host = f'{bucket}.storage.yandexcloud.net'
        self.service = 's3'
        self.endpoint = f'https://{self.host}'

    async def get(
            self,
            path: str = '',
            *,
            params: Optional[Dict[str, Any]] = None
    ) -> Response:
        return await self.request('GET', path=path, params=params)

    async def raw_post(
            self,
            url: str,
            *,
            expected_status: int,
            params: Optional[Dict[str, Any]] = None,
            data: Optional[Dict[str, str]] = None,
            files: Optional[Dict[str, bytes]] = None,
    ) -> Response:
        r = await self.client.post(url, params=params, data=data, files=files)
        if r.status_code == expected_status:
            return r
        else:
            raise RequestError(r)

    async def post(
            self,
            path: str = '',
            *,
            params: Optional[Dict[str, Any]] = None,
            data: Optional[bytes] = None,
            content_type: Optional[str] = None,
    ) -> Response:
        return await self.request('POST',
                                  path=path,
                                  params=params,
                                  data=data,
                                  content_type=content_type)

    async def request(
            self,
            method: Literal['GET', 'POST'],
            *,
            path: str,
            params: Optional[Dict[str, Any]],
            data: Optional[bytes] = None,
            content_type: Optional[str] = None,
    ) -> Response:
        url = URL(f'{self.endpoint}{path}',
                  params=[(k, v) for k, v in sorted((params or {}).items())])
        auth_headers = self._auth_headers(method,
                                          url,
                                          data=data,
                                          content_type=content_type)
        r = await self.client.request(method,
                                      url,
                                      content=data,  # type: ignore
                                      headers=auth_headers,)
        if r.status_code != 200:
            raise RequestError(r)
        return r

    def _auth_headers(
            self,
            method: Literal['GET', 'POST'],
            url: URL,
            *,
            data: Optional[bytes] = None,
            content_type: Optional[str] = None,
    ) -> Dict[str, str]:
        n = utcnow()
        x_amz_date = n.strftime('%Y%m%dT%H%M%SZ')
        date_stamp = n.strftime('%Y%m%d')
        data = data or b''
        content_type = content_type or _CONTENT_TYPE

        # WARNING!
        # order is important here, headers need to be in alphabetical order
        payload_sha256_hash = hashlib.sha256(data).hexdigest()

        headers = {
            'content-md5': b64encode(hashlib.md5(data).digest()).decode(),
            'content-type': content_type,
            'host': self.host,
            'x-amz-content-sha256': payload_sha256_hash,
            'x-amz-date': x_amz_date,
        }

        ctx = dict(
            method=method,
            path=url.path,
            query=url.query.decode(),
            access_key=self.access_key,
            algorithm=_AUTH_ALGORITHM,
            x_amz_date=x_amz_date,
            auth_request=_AWS_AUTH_REQUEST,
            date_stamp=date_stamp,
            payload_sha256_hash=payload_sha256_hash,
            region=self.region,
            service=self.service,
            signed_headers=';'.join(headers),
        )
        ctx.update(credential_scope=_CREDENTIAL_SCOPE.format(**ctx))
        canonical_headers = ''.join(f'{k}:{v}\n' for k, v in headers.items())

        canonical_request = _CANONICAL_REQUEST.format(
            canonical_headers=canonical_headers, **ctx).encode()

        s2s = _STRING_TO_SIGN.format(canonical_request_hash=hashlib.sha256(
            canonical_request).hexdigest(), **ctx)

        key_parts = (
            b'AWS4' + self.secret_key.encode(),
            date_stamp,
            self.region,
            self.service,
            _AWS_AUTH_REQUEST,
            s2s,
        )
        signature: bytes = reduce(_reduce_signature, key_parts)  # type: ignore

        authorization_header = _AUTH_HEADER.format(
            signature=hexlify(signature).decode(), **ctx)
        headers.update({'authorization': authorization_header})
        return headers


def _reduce_signature(key: bytes, msg: str) -> bytes:
    return hmac.new(key, msg.encode(), hashlib.sha256).digest()


class RequestError(RuntimeError):
    def __init__(self, r: Response):
        error_msg = (f'unexpected response from {r.request.method} '
                     f'"{r.request.url}": {r.status_code}')
        super().__init__(error_msg)
        self.response = r
        self.status = r.status_code

    def __str__(self) -> str:
        return f'{self.args[0]}, response:\n{self.response.text}'


s3/_utils.py
import asyncio
from datetime import datetime, timezone
from typing import (
    TYPE_CHECKING,
    Any,
    Awaitable,
    Iterable,
    List,
    Optional
)

__all__ = 'get_config_attr', 'to_unix_s', 'utcnow', 'ManyTasks'

if TYPE_CHECKING:
    from ._types import BaseConfigProtocol


EPOCH = datetime(1970, 1, 1)
EPOCH_TZ = EPOCH.replace(tzinfo=timezone.utc)


def get_config_attr(config: 'BaseConfigProtocol', name: str) -> str:
    try:
        s = getattr(config, name)
    except AttributeError:
        raise TypeError(f'config has not attribute {name}')

    if isinstance(s, str):
        return s
    else:
        raise TypeError(f'config.{name} must be a string not {s.__class__.__name__}')


def to_unix_s(dt: datetime) -> int:
    if dt.utcoffset() is None:
        diff = dt - EPOCH
    else:
        diff = dt - EPOCH_TZ
    return int(round(diff.total_seconds()))


def utcnow() -> datetime:
    return datetime.utcnow().replace(tzinfo=timezone.utc)


class ManyTasks:
    """
    Simply utility to start many tasks without awaiting them,
    then wait for them all to finish
    """

    __slots__ = '_tasks'

    def __init__(self) -> None:
        self._tasks: List[asyncio.Task[Any]] = []

    def add(self, coro: Awaitable[Any], *, name: Optional[str] = None) -> None:
        task = asyncio.create_task(coro, name=name)
        self._tasks.append(task)

    async def finish(self) -> Iterable[Any]:
        return await asyncio.gather(*self._tasks)


s3/s3.py
import base64
import hashlib
import hmac
import json
import mimetypes
import re
from dataclasses import dataclass
from datetime import datetime, timedelta
from itertools import chain
from math import ceil
from typing import (
    Any,
    AsyncIterable,
    Dict,
    List,
    Optional,
    Union
)
from urllib.parse import urlencode
from xml.etree import ElementTree

from pydantic import BaseModel, validator

from ._utils import ManyTasks, to_unix_s, utcnow
from .core import YandexS3Client

__all__ = 'S3Client', 'S3Config', 'S3File'

# rounding of download link expiry time,
# this allows the CDN to efficiently cache download links
expiry_rounding = 100

# removing xmlns="http://s3.amazonaws.com/doc/2006-03-01/"
# from xml makes it much easier to parse
xmlns = 'http://s3.amazonaws.com/doc/2006-03-01/'
xmlns_re = re.compile(f' xmlns="{re.escape(xmlns)}"'.encode())


@dataclass
class S3Config:
    access_key: str
    secret_key: str
    region: str
    s3_bucket: str


class S3File(BaseModel):
    key: str
    last_modified: datetime
    size: int
    e_tag: str
    storage_class: str

    @validator('e_tag')
    def set_ts_now(cls, v: str) -> str:
        return v.strip('"')

    class Config:
        @classmethod
        def alias_generator(cls, string: str) -> str:
            return ''.join(word.capitalize() for word in string.split('_'))


class S3Client:
    __slots__ = '_config', '_client'

    def __init__(
            self,
            access_key: str,
            secret_key: str,
            region: str,
            s3_bucket: str
    ):
        self._config = S3Config(
            access_key=access_key,
            secret_key=secret_key,
            region=region,
            s3_bucket=s3_bucket
        )
        self._client = YandexS3Client(self._config)

    async def list(self, prefix: Optional[str] = None) -> AsyncIterable[S3File]:
        """
        List S3 files with the given prefix.

        https://docs.aws.amazon.com/AmazonS3/latest/API/API_ListObjectsV2.html
        """
        assert prefix is None or not prefix.startswith('/'), \
               'the prefix to filter by should not start with "/"'
        continuation_token = None

        while True:
            params = {
                'list-type': 2,
                'prefix': prefix,
                'continuation-token': continuation_token
            }
            r = await self._client.get(
                params={k: v for k, v in params.items() if v is not None})

            xml_root = ElementTree.fromstring(xmlns_re.sub(b'', r.content))
            for c in xml_root.findall('Contents'):
                yield S3File.parse_obj({v.tag: v.text for v in c})
            if ((t := xml_root.find('IsTruncated')) is not None and
                    t.text == 'false'):
                break

            if (t := xml_root.find('NextContinuationToken')) is not None:
                continuation_token = t.text
            else:
                raise RuntimeError(f'unexpected response from S3: {r.text!r}')

    async def delete(self, *files: Union[str, S3File]) -> List[str]:
        """
        Delete one or more files, based on keys.
        """
        tasks = ManyTasks()
        chunk_size = 1000
        for i in range(0, len(files), chunk_size):
            tasks.add(self._delete_1000_files(*files[i: i + chunk_size]))

        results = await tasks.finish()
        return list(chain(*results))

    async def upload(
            self,
            file_path: str,
            content: bytes,
            *,
            content_type: Optional[str] = None
    ) -> None:
        assert not file_path.startswith('/'), 'file_path must not start with /'
        parts = file_path.rsplit('/', 1)

        if content_type is None:
            content_type, _ = mimetypes.guess_type(file_path)

        d = self.signed_upload_url(
            path=f'{parts[0]}/' if len(parts) > 1 else '',
            filename=parts[-1],
            content_type=content_type or 'application/octet-stream',
            size=len(content),
            expires=datetime.utcnow() + timedelta(minutes=30),
        )
        await self._client.raw_post(
            d['url'],
            expected_status=204,
            data=d['fields'],
            files={'file': content}
        )

    async def download(self, file_path: str) -> bytes:
        assert not file_path.startswith('/'), 'file_path must not start with /'

        url = self.signed_download_url(file_path)
        r = await self._client.client.get(url)
        if r.status_code == 200:
            return r.content
        else:
            raise RuntimeError(f'unexpected response from S3: {r.text!r}')

    async def delete_recursive(self, prefix: Optional[str]) -> List[str]:
        """
        Delete files starting with a specific prefix.
        """
        files = []
        tasks = ManyTasks()
        async for f in self.list(prefix):
            files.append(f)
            if len(files) == 1000:
                tasks.add(self._delete_1000_files(*files))
                files = []

        if files:
            tasks.add(self._delete_1000_files(*files))
        results = await tasks.finish()
        return list(chain(*results))

    async def _delete_1000_files(self, *files: Union[str, S3File]) -> List[str]:
        """
        https://docs.aws.amazon.com/AmazonS3/latest/API/API_DeleteObjects.html
        """
        assert len(files) <= 1000, \
               f'function can delete 1000 files max, not {len(files)}'
        objects = [f"<Object><Key>{to_key(k)}</Key></Object>" for k in files]
        xml = (
            f'<?xml version="1.0" encoding="UTF-8"?>'
            f'<Delete xmlns="{xmlns}">'
            f'{"".join(objects)}'
            f'</Delete>'
        )
        r = await self._client.post(
            '',
            data=xml.encode(),
            params=dict(delete=1),
            content_type='text/xml'
        )
        xml_root = ElementTree.fromstring(xmlns_re.sub(b'', r.content))
        return [k.find('Key').text for k in xml_root]  # type: ignore

    def signed_download_url(
            self,
            path: str,
            version: Optional[str] = None,
            max_age: int = 30
    ) -> str:
        """
        Sign a path to authenticate download.

        The url is valid for between max_age seconds and max_age + expiry_rounding seconds.

        https://docs.aws.amazon.com/AmazonS3/latest/dev/RESTAuthentication.html#RESTAuthenticationQueryStringAuth
        """
        assert not path.startswith('/'), 'path should not start with /'
        min_expires = to_unix_s(utcnow()) + max_age
        expires = int(ceil(min_expires / expiry_rounding) * expiry_rounding)
        to_sign = f'GET\n\n\n{expires}\n/{self._config.s3_bucket}/{path}'
        signature = self._signature(to_sign.encode())
        args = {
            'AWSAccessKeyId': self._config.access_key,
            'Signature': signature,
            'Expires': expires
        }
        if version:
            args['v'] = version
        return f'https://{self._client.host}/{path}?{urlencode(args)}'

    def signed_upload_url(
        self,
        *,
        path: str,
        filename: str,
        content_type: str,
        size: int,
        content_disp: bool = True,
        expires: Optional[datetime] = None,
    ) -> Dict[str, Any]:
        """
        https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-post-example.html
        """
        assert path == '' or path.endswith('/'), \
                'path must be empty or end with "/"'
        assert not path.startswith('/'), 'path must not start with "/"'
        key = path + filename
        policy_conditions = [
            {'bucket': self._config.s3_bucket},
            {'key': key},
            {'content-type': content_type},
            ['content-length-range', size, size],
        ]

        fields = {
            'Key': key,
            'Content-Type': content_type,
            'AWSAccessKeyId': self._config.access_key
        }
        if content_disp:
            disp = {'Content-Disposition': f'attachment; filename="{filename}"'}
            policy_conditions.append(disp)
            fields.update(disp)

        policy = {
            'expiration': f'{expires or utcnow() + timedelta(seconds=60):%Y-%m-%dT%H:%M:%SZ}',
            'conditions': policy_conditions,
        }
        b64_policy: bytes = base64.b64encode(json.dumps(policy).encode())
        fields.update(Policy=b64_policy.decode(),
                      Signature=self._signature(b64_policy))
        return dict(url=f'https://{self._client.host}/', fields=fields)

    def _signature(self, to_sign: bytes) -> str:
        s = hmac.new(self._config.secret_key.encode(),
                     to_sign,
                     hashlib.sha1).digest()
        return base64.b64encode(s).decode()

    async def close(self):
        if not self._client.client.is_closed:
            await self._client.client.aclose()


def to_key(sf: Union[S3File, str]) -> str:
    if isinstance(sf, str):
        return sf
    elif isinstance(sf, S3File):
        return sf.key
    else:
        raise TypeError('must be a string or S3File object')
